sdks/go/pkg/beam/core/runtime/exec/fullvalue.go

```diff
@@ -189,3 +191,173 @@ func ReadAll(rs ReStream) ([]FullValue, error) {
 		ret = append(ret, *elm)
 	}
 }
+
+// decodeReStream is a decode on demand ReStream.
+// Can only produce a single Stream because it consumes the reader.
+// Must not be used for streams that might be re-iterated, causing Open
+// to be called twice.
+type decodeReStream struct {
```

Comment:
Naming nit - does it still make sense to call this a ReStream? ReStream suggests to me that it can be iterated multiple times, and I think that's what the interface itself suggests - https://github.com/apache/beam/blob/48bad7d966a583055669850eb9fb558782f636a8/sdks/go/pkg/beam/core/runtime/exec/fullvalue.go#L60

Reply:
One (unfortunately verbose) option is `singleUseDecodeReStream`?

Reply:
I'm going to drop the "decode" from the restream levels, so singleUseRestream. Simpler to understand that way.

The Stream levels are pretty standard decode streams at least.